source-dropbox/main.go

```diff
@@ -0,0 +1,272 @@
+package main
+
+import (
+	"context"
+	_ "embed"
+	"encoding/json"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/dropbox/dropbox-sdk-go-unofficial/v6/dropbox"
+	"github.com/dropbox/dropbox-sdk-go-unofficial/v6/dropbox/files"
+	"github.com/estuary/connectors/filesource"
+	"github.com/estuary/flow/go/parser"
+	pf "github.com/estuary/flow/go/protocols/flow"
+	log "github.com/sirupsen/logrus"
+)
+
+type config struct {
+	Path        string         `json:"path"`
+	Credentials *Credentials   `json:"credentials"`
+	Parser      *parser.Config `json:"parser"`
+	MatchKeys   string         `json:"matchKeys,omitempty"`
+	Advanced    advancedConfig `json:"advanced"`
+}
+
+type advancedConfig struct {
+	AscendingKeys bool `json:"ascendingKeys,omitempty"`
+}
+
+func (c config) Validate() error {
+	requiredProperties := [][]string{
+		{"Path", c.Path},
+	}
+	for _, req := range requiredProperties {
+		if req[1] == "" {
+			return fmt.Errorf("missing '%s'", req[0])
+		}
+	}
+	return nil
+}
+
+func (c config) DiscoverRoot() string {
+	return c.Path
+}
+
+func (c config) RecommendedName() string {
+	return c.Path[1:]
+}
+
+func (c config) FilesAreMonotonic() bool {
+	return c.Advanced.AscendingKeys
+}
+
+func (c config) ParserConfig() *parser.Config {
+	return c.Parser
+}
+
+func (c config) PathRegex() string {
+	return c.MatchKeys
+}
+
+func newDropboxStore(ctx context.Context, cfg config) (*dropboxStore, error) {
+	token, err := cfg.Credentials.GetAccessToken(ctx)
+	if err != nil {
+		return &dropboxStore{}, err
+	}
+	config := dropbox.Config{
+		Token:    token,
+		LogLevel: dropbox.LogDebug,
+	}
+	client := files.New(config)
+
+	store := dropboxStore{
+		client: client,
+		config: cfg,
+	}
+	if err := store.check(); err != nil {
+		return &dropboxStore{}, err
+	}
+	return &store, nil
+}
+
+type dropboxStore struct {
+	client files.Client
+	config config
+}
+
+// check checks the connection to the Azure Blob Storage container.
+// It returns an error if the container is not found, the account is disabled, or if there is an authorization failure.
+// If the listing is successful, it returns nil.
+func (dbx *dropboxStore) check() error {
+	_, err := dbx.client.ListFolder(&files.ListFolderArg{Path: dbx.config.Path})
+	if err != nil {
+		return fmt.Errorf("failed to list files: %w", err)
+	}
+	return nil
+}
+
+func (dbx *dropboxStore) List(ctx context.Context, query filesource.Query) (filesource.Listing, error) {
+	log.Debug("Listing files in path: ", dbx.config.Path)
```

Comment:
Something isn't working with this file listing logic. I tested this on my dropbox account by adding 10,000 files to a directory and it only captured a single file. There's a lot going on with `ListFolderContinue` and other pagination-looking methods so I wonder if something is being missed there. The lexical ordering of file paths also matters.